File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.swift
let CASE_SENSITIVE: Bool = false;

func compare_string (text1: String, text2: String) -> Bool {
	var nje: Bool = false;
	let len_text1: Int = text1.count;
	let len_text2: Int = text2.count;
	
	if ((len_text1 == 0) && (len_text2 == 0)){
		nje = true;
		return (nje);
	}
	
	if (len_text1 != len_text2){
		return (nje);
	}
	
	var text1_c: String = text1;
	var text2_c: String = text2;
	
	if (CASE_SENSITIVE){
		text1_c.uppercased();
		text2_c.uppercased();
	}
	
	nje = text1_c == text2_c;
	
	return (nje);
}

func first_occurence (needle: String, haystack: String) -> Int{
	let len_needle: Int = needle.count;
	let len_haystack: Int = haystack.count;
	
	var nje: Int = -1;
	
	if (len_needle > len_haystack){
		nje = -1;
	} else if (len_needle == len_haystack){
		nje = (compare_string(text1: needle, text2: haystack)) ? 0 : -1;
	} else {
		let len_searchable: Int = len_haystack - len_needle + 1;
		
		for i in 0..<len_searchable{
			let beg = haystack.index(haystack.startIndex, offsetBy: i);
			let end = haystack.index(haystack.startIndex, offsetBy: i + len_needle);
			var substring: String = String(haystack[beg..<end]);
			
			if (compare_string(text1: needle, text2: substring)){
				nje = i;
				break;
			}
		}
	}
	
	return (nje);
}


let haystack: String = "leetcode";
let needle: String = "od";

print("This is the handy work of Swift.\n\n\n");
print(first_occurence(needle: needle, haystack: haystack));
print("\n\n\n");

File: 28. Find the Index of the First Occurrence in a String/find_the_index_of_the_first_occurence_in_a_string.rs
static CASE_SENSITIVE: bool = false;

fn compare_string (text1: String, text2: String) -> bool {
	let mut nje: bool = false;
	let len_text1: usize = text1.len();
	let len_text2: usize = text2.len();
	
	if ((len_text1 == 0) && (len_text2 == 0)){
		nje = true;
		return (nje);
	}
	
	if (len_text1 != len_text2){
		return (nje);
	}
	
	let mut text1_c: String = text1;
	let mut text2_c: String = text2;
	
	if (CASE_SENSITIVE){
		text1_c.to_uppercase();
		text2_c.to_uppercase();
	}
	
	nje = text1_c == text2_c;
	
	return(nje);
}

fn first_occurence (needle: String, haystack: String) -> isize{
	let len_needle: usize = needle.len();
	let len_haystack: usize = haystack.len();
	
	let mut nje: isize = -1;
	
	if (len_needle > len_haystack){
		nje = -1;
	} else if (len_needle == len_haystack){
		nje = if (compare_string(needle, haystack)) {0} else {-1};
	} else {
		let len_searchable: usize = len_haystack - len_needle +1;
		
		for i in 0..len_searchable{
			let substring: String = haystack[i..(i + len_needle)].to_string();
			
			if (compare_string(needle.clone(), substring.clone())){
				nje = i as isize;
				break;
			}
		}
	}
	
	return (nje);
}

fn main(){
	let haystack: String = "leetcode".to_string();
	let needle: String = "ode".to_string();
	
	println!("This is the handy work of Rust.\n\n\n{}\n\n\n", first_occurence(needle.clone(), haystack.clone()));
}
